### PR TITLE
[IMP]: Make source document clickable in chatter.

### PR DESCRIPTION
**Pull request by atp-odoo**
Task: https://www.odoo.com/web?#id=19637&view_type=form&model=project.task&action=333&active_id=131&menu_id=4720
Pad: https://pad.odoo.com/p/r.1984de291877d0e85fa71f32f3ef1bf6

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:
## 

I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by tivisse** [comment]
Strange way to divide your commits.

File: addons/sale_stock/sale_stock.py [hunk 1]
```diff
@@ -271,6 +271,20 @@ def _run_move_create(self):
             vals.update({'sequence': self.sale_line_id.sequence})
         return vals
 
+    @api.cr_uid_context
+    def create(self, cr, uid, vals, context=None):
```
Comment (tivisse) [inline addons/sale_stock/sale_stock.py]: Why writing it old api ? We do not need any mapping here.

Why not use @api.model ?

File: addons/sale_stock/sale_stock.py [hunk 2]
```diff
@@ -291,6 +305,25 @@ def action_done(self):
             line.qty_delivered = line._get_delivered_qty()
         return result
 
+    @api.multi
+    def assign_picking(self):
```
Comment (tivisse) [inline addons/sale_stock/sale_stock.py]: I bet you can find the picking easily without doing and search like this...

File: addons/sale_stock/sale_stock.py [hunk 3]
```diff
@@ -307,6 +340,16 @@ def _compute_sale_id(self):
 
     sale_id = fields.Many2one(comodel_name='sale.order', string="Sale Order", compute='_compute_sale_id')
 
+    @api.multi
+    def _create_backorder(self, backorder_moves=[]):
+        res = super(StockPicking, self)._create_backorder(backorder_moves)
+        for picking in self.filtered(lambda pick: pick.picking_type_id.code == 'outgoing'):
```
Comment (tivisse) [inline addons/sale_stock/sale_stock.py]: Why are you forced to do this ?

File: addons/sale_stock/sale_stock.py (hunk 2 above)
Comment (ptr-odoo) [inline addons/sale_stock/sale_stock.py]: In the stock module  `def assign_picking` is create picking by using search method.
when we install sale module we can not directly get picking. that why we override method in sale_stock.
I think currently we can not fix code in stock module.

File: addons/sale_stock/sale_stock.py (hunk 3 above)
Comment (atp-odoo) [inline addons/sale_stock/sale_stock.py]: purpose is we need to set link of sale order in backorder of picking.but when we do partial picking and backorder is created we cannot identity any reference from so. so we need to search first all picking for outgoing, 
then find backorder of that picking and make order reference on back order. 

that why we need to search by outgoing picking here.

Comment (tivisse) [inline addons/sale_stock/sale_stock.py]: Ok thank you :)

File: addons/sale_stock/sale_stock.py (hunk 2 above)
Comment (tivisse) [inline addons/sale_stock/sale_stock.py]: The original method  assigns an existing picking or a new one on the move. So I think that we can reuse it on the override. Do not modify your branch, I'm working on it.

Thank you

File: addons/purchase/purchase.py [hunk 4]
```diff
@@ -1078,7 +1087,8 @@ class MailComposeMessage(models.Model):
 
     @api.multi
     def send_mail(self, auto_commit=False):
-        if self._context.get('default_model') == 'purchase.order' and self._context.get('default_res_id'):
+        compose_internal = self.filtered('subtype_id.internal')
```
Comment (tivisse) [inline addons/purchase/purchase.py]: Why did you do this ?

File: addons/purchase/stock.py [hunk 5]
```diff
@@ -21,6 +21,17 @@ def _prepare_values_extra_move(self, cr, uid, op, product, remaining_qty, contex
                 break
         return res
 
+    @api.cr_uid_ids_context
```
Comment (tivisse) [inline addons/purchase/stock.py]: I still don't understand why you write this in old api

File: addons/purchase/purchase.py (hunk 4 above)
Comment (atp-odoo) [inline addons/purchase/purchase.py]: we used  "message_post_with_view" method to render template and send mail which finally use send_mail method and this send_mail method is already overwriten in purchase,sale etc which update draft state. to send..

see send mail method in purchase.
https://github.com/odoo/odoo/blob/master/addons/purchase/purchase.py#L1064

problem is, when po is generated from procurement , this "make_po" method in purchase  will first find all draft state Po if no draft created then create new po. but in our case when send mail called from message_post_with_tempalte, this method update draft po to send . and
everytime make_po method call new po eventhough we have already related purchase order created.  so we avoid to update state when we have "internal subtype".

see make_po method.
https://github.com/odoo/odoo/pull/11612/files#diff-fe8587929089e5fdb6c75687011b1c7eR926

File: addons/sale_stock/sale_stock.py (hunk 1 above)
Comment (atp-odoo) [inline addons/sale_stock/sale_stock.py]: You are right.. @api.model should work. but at that time not sure, sale is not migration 
and this create method using @api.model generated error. so we used that . 
but i tested now @api.model work.. 
Thanks

File: addons/purchase/purchase.py (hunk 4 above)
Comment (tivisse) [inline addons/purchase/purchase.py]: Thank you for your answer

File: addons/sale_stock/sale_stock.py [hunk 6]
```diff
@@ -271,6 +271,17 @@ def _run_move_create(self):
             vals.update({'sequence': self.sale_line_id.sequence})
         return vals
 
+    @api.model
+    def create(self, vals):
+        procurement = super(ProcurementOrder, self).create(vals)
+        move = procurement.move_dest_id
+        if move.group_id:
+            order = self.env['sale.order'].search([('procurement_group_id', '=', move.group_id.id)])
```
Comment (tivisse) [inline addons/sale_stock/sale_stock.py]: I don't like the fact that the child record is looking for his father, in order to send himself a message. I'll rework this part

**Pull request merged**